src/libmongoc/tests/test-mcd-rpc.c

```diff
@@ -0,0 +1,3165 @@
+}
+
+static int64_t
+_int64_from_le (const void *data)
+{
+   BSON_ASSERT_PARAM (data);
+   return bson_iter_int64_unsafe (&(bson_iter_t){.raw = data});
+}
+
+
+static void
+test_rpc_message_from_data_op_compressed_valid (void)
+{
+   uint8_t data[] = {TEST_DATA_OP_COMPRESSED};
+   const size_t data_len = sizeof (data);
+
+   // Valid test input data.
+   {
+      const void *data_end = NULL;
+      mcd_rpc_message *const rpc =
+         mcd_rpc_message_from_data (data, data_len, &data_end);
+
+      ASSERT_RPC_MESSAGE_RESULT (rpc, data, data_end, data_len);
+      ASSERT (bson_in_range_unsigned (int32_t, data_len));
+      ASSERT_CMPINT32 (
+         mcd_rpc_header_get_message_length (rpc), ==, (int32_t) data_len);
+      ASSERT_CMPINT32 (mcd_rpc_header_get_request_id (rpc), ==, 16909060);
+      ASSERT_CMPINT32 (mcd_rpc_header_get_response_to (rpc), ==, 84281096);
+      ASSERT_CMPINT32 (
+         mcd_rpc_header_get_op_code (rpc), ==, MONGOC_OP_CODE_COMPRESSED);
+
+      ASSERT_CMPINT32 (mcd_rpc_op_compressed_get_original_opcode (rpc),
+                       ==,
+                       MONGOC_OP_CODE_MSG);
+
+      ASSERT_CMPINT32 (
+         mcd_rpc_op_compressed_get_uncompressed_size (rpc), ==, 18);
+
+      ASSERT_CMPUINT (mcd_rpc_op_compressed_get_compressor_id (rpc),
+                      ==,
+                      MONGOC_OP_COMPRESSED_COMPRESSOR_ID_NOOP);
+
+      const uint8_t *const compressed_message =
+         mcd_rpc_op_compressed_get_compressed_message (rpc);
+      ASSERT_CMPSIZE_T ((size_t) (compressed_message - data), ==, 25u);
+
+      ASSERT_CMPSIZE_T (
+         mcd_rpc_op_compressed_get_compressed_message_length (rpc), ==, 18u);
+
+      mcd_rpc_message_destroy (rpc);
+   }
+
+   // Test that compressorId is being parsed correctly.
+   {
+      data[24] = MONGOC_OP_COMPRESSED_COMPRESSOR_ID_SNAPPY;
+
+      mcd_rpc_message *const rpc =
+         mcd_rpc_message_from_data (data, data_len, NULL);
+      ASSERT (rpc);
+      ASSERT_CMPUINT (mcd_rpc_op_compressed_get_compressor_id (rpc),
+                      ==,
+                      MONGOC_OP_COMPRESSED_COMPRESSOR_ID_SNAPPY);
+      mcd_rpc_message_destroy (rpc);
+
+      data[24] = MONGOC_OP_COMPRESSED_COMPRESSOR_ID_NOOP;
+   }
+}
+
+
+static void
+_test_rpc_message_from_data_op_msg_valid (uint8_t *data,
+                                          size_t data_len,
+                                          void (*test) (const uint8_t *data,
+                                                        size_t data_len,
+                                                        bool with_checksum))
+{
+   ASSERT_WITH_MSG (
+      (data[16] & MONGOC_OP_MSG_FLAG_CHECKSUM_PRESENT) != 0,
+      "test input data did not set MONGOC_OP_MSG_FLAG_CHECKSUM_PRESENT");
+
+   // Test with and without the optional checksum by temporarily modifying the
+   // data and data length accordingly.
+   {
+      data[0] = (uint8_t) (data[0] - 4u); // Reduce messageLength.
```

Comment:
Given this test is highly dependent on the test data it is being used with, and the test data (for all three situations being tested against: Kind 0, Kind 1 Single, and Kind 1 Multiple) being used all have `messageLength > 4` (this is a _valid_ data test after all), I consider this suggestion out-of-scope for this function. This function's sole purpose is to avoid code duplication for the "with checksum" and "without checksum" test cases. A valid `messageLength` must have a value of at least 16 (for all 4 header fields), so overflow is not a concern here.

Reply:
Agreed. Handling this case seems unnecessary.